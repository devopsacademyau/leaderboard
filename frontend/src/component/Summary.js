import React, { Fragment } from "react"

const Summary = ({ charts, config, bot }) => {
  if (!charts || charts.length === 0) {
    return null
  }
  const chartsByTitle = charts.reduce((obj, chart) => {
    obj[chart.title] = chart
    return obj
  }, {})
  // TODO(#7): Make charts generic
  const pr = chartsByTitle["Merged Pull Requests"] || null
  const open = chartsByTitle["Merged or Open Pull Requests"] || null
  const review = chartsByTitle["Code Reviews"] || null
  const comment = chartsByTitle["Code Review Comments"] || null
  return (
    <Fragment>
      <PrSummary pr={pr} open={open} config={config} />
      <ReviewSummary {...{ review, comment, bot }} />
      <Note config={config} />
    </Fragment>
  )
}

const P = ({ children }) => <p className="lh-copy measure">{children}</p>
const Code = ({ children }) => (
  <code className="f7 fs-normal ba b--black-20 br1 ph1">{children}</code>
)

const PrSummary = ({ pr, open, config }) => {
  if (!pr || !open) {
    return null
  }
  const repoUrl = config.repository.url
  const path = "/pulls"
  const createdAfter = config.createdAfter.substring(0, 10)
  const queryOpen = "?q=is%3Apr+is%3Aopen+created%3A>" + createdAfter
  const queryMerged = "?q=is%3Apr+is%3Amerged+created%3A>" + createdAfter
  const openUrl = `${repoUrl}${path}${queryOpen}`
  const mergedUrl = `${repoUrl}${path}${queryMerged}`
  return (
    <P>
      <b>
        {pr.points.length} people{" "}
        <Link url={mergedUrl}>merged {pr.totalCount} Pull Requests</Link>
      </b>
      , with a highest individual contribution of {pr.maxCount}. There are an
      additional{" "}
      <Link url={openUrl}>{open.totalCount - pr.totalCount} open PRs</Link> in
      the works. The highest individual contribution of open and merged PRs is{" "}
      {open.maxCount}.
    </P>
  )
}

const ReviewSummary = ({ review, comment, bot }) =>
  review &&
  comment && (
    <P>
      <b>{review.points.length} authors</b> created
      <b>{` ${review.totalCount} ${review.title}`}</b> and
      <b>{` ${comment.totalCount} ${comment.title}`}</b>.{" "}
      <Link url={bot.url} ttc>
        <img src={bot.avatarUrl} alt="avatar" className="h1 w1 ph1" />
        {bot.login}
      </Link>{" "}
      made {bot.comments} comments.
    </P>
  )

const Note = ({ config }) => (
  <p className="lh-copy measure-wide f6 i">
    PRs have been filtered by labels matching <Code>{config.labelGlob}</Code>{" "}
    and creation date after <Code>{config.createdAfter.substring(0, 10)}</Code>
  </p>
)

const Link = ({ url, children, ttc }) => {
  const classNamePrefix = "link black-80 hover-blue underline-hover"
  const className = `${classNamePrefix} ${ttc ? "ttc nowrap" : "b"}`
  const rel = "noopener noreferrer"
  const target = "_blank"
  return (
    <a href={url} target={target} rel={rel} className={className}>
      {children}
    </a>
  )
}

export default Summary
