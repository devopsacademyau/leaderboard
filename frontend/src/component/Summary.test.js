import React from "react"
import Summary from "./Summary"
import renderer from "react-test-renderer"

test("matches snapshot for empty charts array", () => {
  const tree = renderer.create(<Summary charts={[]} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test("matches snapshot charts", () => {
  const charts = [
    {
      title: "Merged Pull Requests",
      maxCount: 4,
      totalCount: 16,
      points: [1, 2, 3, 4, 5] // only length of points matters
    },
    {
      title: "Merged or Open Pull Requests",
      maxCount: 5,
      totalCount: 73,
      points: [1, 2, 3, 4, 5, 6, 7]
    },
    {
      title: "Code Reviews",
      maxCount: 47,
      totalCount: 382,
      points: [1, 2, 3]
    },
    {
      title: "Code Review Comments",
      maxCount: 121,
      totalCount: 852,
      points: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    }
  ]
  const config = {
    labelGlob: "lab.*",
    botName: "golangcibot",
    createdAfter: "2019-05-15T00:00:00Z",
    repository: {
      name: "academy",
      owner: "devopsacademyau",
      url: "https://github.com/devopsacademyau/academy"
    }
  }
  const bot = {
    login: "golangcibot",
    url: "https://github.com/golangcibot",
    avatarUrl: "https://avatars1.githubusercontent.com/u/42910462?v=4",
    comments: 2000
  }
  const tree = renderer
    .create(<Summary charts={charts} config={config} bot={bot} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("matches snapshot charts with unknown title", () => {
  const charts = [
    {
      title: "UNKNOWN CHART TITLE",
      maxCount: 121,
      totalCount: 852,
      points: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    }
  ]
  const config = {
    labelGlob: "lab.*",
    botName: "golangcibot",
    createdAfter: "2019-05-15T00:00:00Z",
    repository: {
      name: "academy",
      owner: "devopsacademyau",
      url: "https://github.com/devopsacademyau/academy"
    }
  }

  const tree = renderer
    .create(<Summary charts={charts} config={config} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
