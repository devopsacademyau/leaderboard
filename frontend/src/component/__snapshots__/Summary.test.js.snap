// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot charts 1`] = `
Array [
  <p
    className="lh-copy measure"
  >
    <b>
      5
       people
       
      <a
        className="link black-80 hover-blue underline-hover b"
        href="https://github.com/devopsacademyau/academy/pulls?q=is%3Apr+is%3Amerged+created%3A>2019-05-15"
        rel="noopener noreferrer"
        target="_blank"
      >
        merged 
        16
         Pull Requests
      </a>
    </b>
    , with a highest individual contribution of 
    4
    . There are an additional
     
    <a
      className="link black-80 hover-blue underline-hover b"
      href="https://github.com/devopsacademyau/academy/pulls?q=is%3Apr+is%3Aopen+created%3A>2019-05-15"
      rel="noopener noreferrer"
      target="_blank"
    >
      57
       open PRs
    </a>
     in the works. The highest individual contribution of open and merged PRs is
     
    5
    .
  </p>,
  <p
    className="lh-copy measure"
  >
    <b>
      3
       authors
    </b>
     created
    <b>
       382 Code Reviews
    </b>
     and
    <b>
       852 Code Review Comments
    </b>
    .
     
    <a
      className="link black-80 hover-blue underline-hover ttc nowrap"
      href="https://github.com/golangcibot"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="avatar"
        className="h1 w1 ph1"
        src="https://avatars1.githubusercontent.com/u/42910462?v=4"
      />
      golangcibot
    </a>
     
    made 
    2000
     comments.
  </p>,
  <p
    className="lh-copy measure-wide f6 i"
  >
    PRs have been filtered by labels matching 
    <code
      className="f7 fs-normal ba b--black-20 br1 ph1"
    >
      lab.*
    </code>
     
    and creation date after 
    <code
      className="f7 fs-normal ba b--black-20 br1 ph1"
    >
      2019-05-15
    </code>
  </p>,
]
`;

exports[`matches snapshot charts with unknown title 1`] = `
<p
  className="lh-copy measure-wide f6 i"
>
  PRs have been filtered by labels matching 
  <code
    className="f7 fs-normal ba b--black-20 br1 ph1"
  >
    lab.*
  </code>
   
  and creation date after 
  <code
    className="f7 fs-normal ba b--black-20 br1 ph1"
  >
    2019-05-15
  </code>
</p>
`;

exports[`matches snapshot for empty charts array 1`] = `null`;
